pgjdbc/src/main/java/org/postgresql/core/BaseConnection.java

```diff
@@ -228,4 +232,227 @@ CachedQuery createQuery(String sql, boolean escapeProcessing, boolean isParamete
+   * the request and no dependency on local state that crosses request
+   * boundaries. Committed database state is not local.
+   *
+   * @throws SQLException if an error occurs
+   * @implSpec The default implementation is a no-op.
+   * @apiNote This method is to be used by Connection pooling managers.
+   * <p>
+   * The pooling manager should call {@code beginRequest} on the underlying connection
+   * prior to returning a connection to the caller.
+   * <p>
+   * The pooling manager does not need to call {@code beginRequest} if:
+   * <ul>
+   * <li>The connection pool caches {@code PooledConnection} objects</li>
+   * <li>Returns a logical connection handle when {@code getConnection} is
+   * called by the application</li>
+   * <li>The logical {@code Connection} is closed by calling
+   * {@code Connection.close} prior to returning the {@code PooledConnection}
+   * to the cache.</li>
+   * </ul>
+   * @see endRequest
+   * @see PooledConnection
+   * @since 9
+   */
+  @Override
+  default void beginRequest() throws SQLException {
+    throw new SQLFeatureNotSupportedException();
+  }
+
+  /**
+   * Hints to the driver that a request, an independent unit of work,
+   * has completed. Calls to {@code beginRequest}
+   * and {@code endRequest} are not nested. Multiple
+   * calls to {@code endRequest} without an intervening call to {@code beginRequest}
+   * is not an error. The first {@code endRequest} call
+   * marks the request completed and subsequent calls are treated as
+   * a no-op. If {@code endRequest} is called without an initial call to
+   * {@code beginRequest} is a no-op.
+   * <p>
+   * The exact behavior of this method is vendor specific. In particular
+   * implementations may detect conditions that indicate dependence on
+   * other work such as an open transaction. It is recommended though not
+   * required that implementations throw a {@code SQLException} if there is an active
+   * transaction and {@code endRequest} is called.
+   *
+   * @throws SQLException if an error occurs
+   * @implSpec The default implementation is a no-op.
+   * @apiNote This method is to be used by Connection pooling managers.
+   * <p>
+   * The pooling manager should call {@code endRequest} on the underlying connection
+   * when the applications returns the connection back to the connection pool.
+   * <p>
+   * The pooling manager does not need to call {@code endRequest} if:
+   * <ul>
+   * <li>The connection pool caches {@code PooledConnection} objects</li>
+   * <li>Returns a logical connection handle when {@code getConnection} is
+   * called by the application</li>
+   * <li>The logical {@code Connection} is closed by calling
+   * {@code Connection.close} prior to returning the {@code PooledConnection}
+   * to the cache.</li>
+   * </ul>
+   * @see beginRequest
+   * @see PooledConnection
+   * @since 9
+   */
+  @Override
+  default void endRequest() throws SQLException {
+    throw new SQLFeatureNotSupportedException();
```

Comment:
Make this a noop instead - connection pools won't expect a SQLFeatureNotSupportedException here. IFF supported, this will be called for every transaction.